pyo3-macros-backend/src/intopydict.rs

```diff
@@ -0,0 +1,366 @@
+          //     }
+
+          //     match val.as_ref() {
+          //         Pyo3Type::Primitive => todo!(),
+          //         Pyo3Type::NonPrimitive => todo!(),
+          //         Pyo3Type::CollectionSing(_) => todo!(),
+          //         Pyo3Type::Map(_, _) => todo!(),
+          //     }
+          // }
+    };
+}
+
+fn handle_single_collection_code_gen(
+    py_type: &Pyo3Type,
+    ident: &str,
+    non_class_ident: &str,
+    counter: usize,
+) -> TokenStream {
+    let curr_pylist: TokenStream = format!("pylist{}{}", counter, non_class_ident)
+        .parse()
+        .unwrap();
+    let next_pylist: TokenStream = format!("pylist{}{}", counter + 1, non_class_ident)
+        .parse()
+        .unwrap();
+    let ident_tok: TokenStream = ident.parse().unwrap();
+    match py_type {
+        Pyo3Type::Primitive => {
+            quote! {
+                let mut #curr_pylist = pyo3::types::PyList::empty(py);
+                for i in #ident_tok.into_iter() {
+                    #curr_pylist.append(i).expect("Bad element in set_item");
+                };
+            }
+        }
+        Pyo3Type::NonPrimitive => {
+            quote! {
+                let mut #curr_pylist = pyo3::types::PyList::empty(py);
+                for i in #ident_tok.into_iter() {
+                    #curr_pylist.append(i.into_py_dict(py)).expect("Bad element in set_item");
+                };
+            }
+        }
+        Pyo3Type::CollectionSing(coll) => {
+            let body =
+                handle_single_collection_code_gen(coll.as_ref(), "i", non_class_ident, counter + 1);
+            quote! {
+                let mut #curr_pylist = pyo3::types::PyList::empty(py);
+                for i in #ident_tok.into_iter(){
+                    #body
+                    #curr_pylist.append(#next_pylist).expect("Bad element in set_item");
+                };
+            }
+        }
+    }
+}
+
+pub fn parse_generics(generics: &Generics) -> String {
+    if !generics.params.is_empty() {
+        let mut generics_parsed = "<".to_string();
+
+        for param in &generics.params {
+            match param {
+                syn::GenericParam::Lifetime(lt) => {
+                    generics_parsed += ("'".to_string() + &lt.lifetime.ident.to_string()).as_str()
+                }
+                syn::GenericParam::Type(generic_type) => {
+                    generics_parsed += generic_type.ident.to_string().as_str()
+                }
+                syn::GenericParam::Const(const_type) => {
+                    generics_parsed +=
+                        ("const".to_string() + const_type.ident.to_string().as_str()).as_str()
+                }
+            }
+
+            generics_parsed += ",";
+        }
+
+        generics_parsed = generics_parsed[0..generics_parsed.len() - 1].to_string();
+        generics_parsed += ">";
+        generics_parsed
+    } else {
+        String::new()
+    }
+}
```

Comment:
[Syn has a function for doing this](https://docs.rs/syn/latest/syn/struct.Generics.html#method.split_for_impl).

Reply:
Thank you! Ill keep that in mind